### PR TITLE
CUDA and redesign

### PR DESCRIPTION
**Pull request by Marco-Christiani**

### CODE REVIEW

File: src/ndtensor.zig [hunk 1]
```diff
@@ -95,6 +95,33 @@ pub fn NDTensor(comptime T: type) type {
...
+            self.* = Self{
+                .data = try dtype.zeros(shape, device),
+                .grad = if (requires_grad) try dtype.zeros(shape, device) else null,
+                .requires_grad = requires_grad,
+                .device = device,
+            };
+            return self;
+        }
+
+        pub fn getShape(self: Self) []const usize {
```
Comment (andrewCodeDev) [inline src/ndtensor.zig:109]: I had to prefix these with "get" because we are getting name collisions otherwise.

Comment (Marco-Christiani) [reply src/ndtensor.zig:109]: lgtm

File: src/ndtensor.zig [hunk 2]
```diff
@@ -663,64 +775,63 @@ pub fn NDTensor(comptime T: type) type {
...
+                fn dotBwImpl(_self: Self) !void {
                     const children = _self.children orelse return error.NoChildren;
                     var a = children[0];
                     var b = children[1];
-                    const grad_a = try b.data.mul(_self.grad.?, bw_allocator);
-                    const grad_b = try a.data.mul(_self.grad.?, bw_allocator);
-                    _ = try a.grad.?._add(grad_a);
-                    _ = try b.grad.?._add(grad_b);
+                    const gscalar: *const T = @ptrCast(_self.grad.?.data.ptr);
+                    _self.device.blas.axpy(T, a.getData(), a.grad.?.data, gscalar);
```
Comment (Marco-Christiani) [inline src/ndtensor.zig:824]: The math aint math'n here

Comment (andrewCodeDev) [reply src/ndtensor.zig]: It should be math'n again.

File: src/cuda/nn/nll_loss.cu [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+) {
+  cudaStream_t stream;
+  CUDNN_ASSERT(cudnnGetStream(static_cast<cudnnHandle_t>(cudnn_handle), &stream));
+  const auto _src = static_cast<T*>(src);
+  const auto _rdx = static_cast<T*>(rdx);
+  __nll_loss_1D_index_reduce_kernel<T><<<1,1,0,stream>>>(_src, trg, n, _rdx, ignore_idx, reduxop);
+  return;
+}
+
+extern "C" void nll_loss_1D_index_reduce(
```
Comment (andrewCodeDev) [inline src/cuda/nn/nll_loss.cu:44]: I need to add a `DtoH` memcpy here and we can just return the float directly. I think synchronization is fine here because if you're printing loss then you need the value in hand.

File: src/ndarray/shape.zig [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+}
+
+pub fn equal(a: Shape, b: Shape) bool {
+    return @reduce(ReduceOp.And, a.simd() == b.simd());
+}
+
+// I'd like a better name for this. Compatible for what, exactly?
+// It's definitely not "equal", but we're seeing if we disagree
+// at very specific moments
+pub fn compatible(a: Shape, b: Shape) bool {
```
Comment (Marco-Christiani) [inline src/ndarray/shape.zig:180]: This is a broadcast check, `can_broadcast` or similar would be a good name right?

File: src/ndarray/shape.zig [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+pub fn head(self: anytype, amount: u64) MatchedSlice(@TypeOf(&self.buffer)) {
+    return self.buffer[0..amount];
+}
+
+pub fn tail(self: anytype, amount: u64) MatchedSlice(@TypeOf(&self.buffer)) {
+    return self.buffer[self.len - amount .. self.len];
+}
+
+pub fn crop(self: anytype, lhs: u64, rhs: u64) MatchedSlice(@TypeOf(&self.buffer)) {
```
Comment (Marco-Christiani) [inline src/ndarray/shape.zig:263]: I think we should rename this to slice. slicing is typically the term used when extracting a range of elements so I feel like this would be more intuitive. We could either remove the above slice() impl altogether (since its just slice(0, n)) or rename it to first_n (but this feels unnecessary to have its own method imo).

If they're separate bc you're really trying to optimize for performance, then we can add a fast path to slice. If we are using slice to get the first n elements, then lhs==0 will be known at comptime. This should give us a zero-cost fast path.

File: src/ndarray/shape.zig [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+    return if (lhs < self.len and rhs < self.len)
+        self.buffer[lhs..@max(lhs, self.len - rhs)]
+    else
+        &.{};
+}
+
+fn _suffix_scan(self: Shape) SimdType {
```
Comment (Marco-Christiani) [inline src/ndarray/shape.zig:270]: for small shapes, which are common, this will likely be slow. that was what I found last time I did this, looks good though.

**Comment by Marco-Christiani** [comment]
@andrewCodeDev I say send it, its just going to keep diverging at this point and the examples work. I need to start bringing the other branches up to date with unstable. Also, rebase makes sense, I prepped everything.
